File: python/panchang/core/sun.py
"""Sunrise, sunset, and solar calculations for Lilavati.

Uses Rust _core extension which wraps Swiss Ephemeris swe_rise_trans()
for precise geometric sunrise/sunset with Hindu rising method.
"""

from __future__ import annotations

from datetime import datetime
from zoneinfo import ZoneInfo

from panchang._core import py_compute_sun_data, py_init
from panchang.core.ephemeris import EphemerisEngine, jd_to_datetime
from panchang.types import Location, SunData


def _tz_offset_for_date(tz_name: str, year: int, month: int, day: int) -> int:
    """Get UTC offset in seconds for a specific date (handles DST)."""
    tz = ZoneInfo(tz_name)
    dt = datetime(year, month, day, 12, 0, 0, tzinfo=tz)
    return int(dt.utcoffset().total_seconds())


def _get_sun_raw(dt: datetime, location: Location) -> dict:
    """Call Rust core to get raw sun data."""
    py_init(None)
    tz = ZoneInfo(location.tz)

    if dt.tzinfo is None:
        local_dt = dt.replace(tzinfo=tz)
    else:
        local_dt = dt.astimezone(tz)

    utc_offset = _tz_offset_for_date(location.tz, local_dt.year, local_dt.month, local_dt.day)

    return py_compute_sun_data(
        local_dt.year,
        local_dt.month,
        local_dt.day,
        location.lat,
        location.lng,
        location.altitude,
        utc_offset,
    )


def compute_sunrise(
    dt: datetime,
    location: Location,
    engine: EphemerisEngine | None = None,
) -> datetime:
    """Compute sunrise for a given date and location.

    Returns sunrise as a timezone-aware datetime in the location's timezone.
    """
    raw = _get_sun_raw(dt, location)
    sunrise_utc = jd_to_datetime(raw["sunrise_jd"])
    return sunrise_utc.astimezone(ZoneInfo(location.tz))


def compute_sunset(
    dt: datetime,
    location: Location,
    engine: EphemerisEngine | None = None,
) -> datetime:
    """Compute sunset for a given date and location.

    Returns sunset as a timezone-aware datetime in the location's timezone.
    """
    raw = _get_sun_raw(dt, location)
    sunset_utc = jd_to_datetime(raw["sunset_jd"])
    return sunset_utc.astimezone(ZoneInfo(location.tz))


def compute_sun_data(
    dt: datetime,
    location: Location,
    engine: EphemerisEngine | None = None,
) -> SunData:
    """Compute complete solar data for a date and location.

    Returns sunrise, sunset, and day duration.
    """
    raw = _get_sun_raw(dt, location)
    tz = ZoneInfo(location.tz)

    sunrise = jd_to_datetime(raw["sunrise_jd"]).astimezone(tz)
    sunset = jd_to_datetime(raw["sunset_jd"]).astimezone(tz)

    return SunData(
        sunrise=sunrise,
        sunset=sunset,
        day_duration_hours=raw["day_duration_hours"],
        sunrise_jd=raw["sunrise_jd"],
        sunset_jd=raw["sunset_jd"],
    )

File: crates/panchang-core/src/sun.rs
//! Sunrise, sunset, and solar calculations.
//!
//! Uses Swiss Ephemeris `swe_rise_trans()` with Hindu rising method:
//! disc center at horizon with Bhāratīya atmospheric model.

use crate::ffi;
use crate::types::SunData;

/// Rise/set flags: Hindu rising method.
const RISE_FLAGS: i32 = ffi::SE_BIT_HINDU_RISING;

/// Compute sunrise Julian Day for a given date and location.
///
/// `jd_start` should be the JD of local midnight (start of day).
/// `swe_rise_trans` searches FORWARD from the given JD.
pub fn sunrise_jd(jd_start: f64, lat: f64, lng: f64, alt: f64) -> f64 {
    let mut geopos = [lng, lat, alt];
    let mut tret = [0.0f64; 10];
    let mut serr = [0i8; ffi::SE_ERR_LEN];

    unsafe {
        ffi::swe_rise_trans(
            jd_start,
            ffi::SE_SUN,
            std::ptr::null(),
            ffi::SEFLG_MOSEPH,
            ffi::SE_CALC_RISE | RISE_FLAGS,
            geopos.as_mut_ptr(),
            0.0, // atpress (0 = default)
            0.0, // attemp (0 = default)
            tret.as_mut_ptr(),
            serr.as_mut_ptr(),
        );
    }
    tret[0]
}

/// Compute sunset Julian Day for a given date and location.
///
/// `jd_noon` should be the JD of local noon. Sunset always comes after noon.
pub fn sunset_jd(jd_noon: f64, lat: f64, lng: f64, alt: f64) -> f64 {
    let mut geopos = [lng, lat, alt];
    let mut tret = [0.0f64; 10];
    let mut serr = [0i8; ffi::SE_ERR_LEN];

    unsafe {
        ffi::swe_rise_trans(
            jd_noon,
            ffi::SE_SUN,
            std::ptr::null(),
            ffi::SEFLG_MOSEPH,
            ffi::SE_CALC_SET | RISE_FLAGS,
            geopos.as_mut_ptr(),
            0.0,
            0.0,
            tret.as_mut_ptr(),
            serr.as_mut_ptr(),
        );
    }
    tret[0]
}

/// Compute sunrise, sunset, and day duration for a date and location.
///
/// `utc_offset_seconds`: timezone offset from UTC in seconds (e.g. IST = 19800).
pub fn compute_sun_data(
    year: i32,
    month: u32,
    day: u32,
    lat: f64,
    lng: f64,
    alt: f64,
    utc_offset_seconds: i32,
) -> SunData {
    crate::ephemeris::init(None);

    let jd_midnight = crate::julian::midnight_jd(year, month, day, utc_offset_seconds);
    let jd_noon = jd_midnight + 0.5;

    let rise = sunrise_jd(jd_midnight, lat, lng, alt);
    let set = sunset_jd(jd_noon, lat, lng, alt);
    let duration = (set - rise) * 24.0;

    SunData {
        sunrise_jd: rise,
        sunset_jd: set,
        day_duration_hours: (duration * 10000.0).round() / 10000.0,
    }
}
